src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+			int num_children = temp_dom->do_target_nr;
+			int last = num_children - 1;
+			struct pool_target *start = &(temp_dom->do_targets[0]);
+			struct pool_target *end = &(temp_dom->do_targets[last]);
+
+			if ((start <= (*target)) && ((*target) <= end)) {
+				current_dom = temp_dom;
+				setbit(dom_used, (index + child_pos));
+				break;
+			}
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * This function handles getting the initial layout for the object as well as
+ * determining if there are targets that are unavailable.
+ *
+ * \param[in] mmap		The placement map used for this placement.
+ * \param[in] mop		The layout group size and count.
+ * \param[in] md		Object metadata.
+ * \param[out] layout		This will contain the layout for the object
+ * \param[out] remap_list	This will contain the targets that need to
+ *				be rebuilt and in the case of rebuild, may be
+ *				returned during the rebuild process.
+ *
+ * \return			An error code determining if the function
+ *				succeeded (0) or failed.
+ */
+static int
+get_object_layout(struct pl_mapless_map *mmap, struct pl_obj_layout *layout,
+		  struct mpls_obj_placement *mop, d_list_t *remap_list,
+		  struct daos_obj_md *md)
+{
+	struct pool_target      *target;
+	uint8_t                 *dom_used;
+	uint32_t                *used_targets;
+	uint64_t                key;
+	daos_obj_id_t		oid;
+	struct pool_domain      *root;
+	int i, j, k, rc;
+
+	/* Set the pool map version */
+	layout->ol_ver = pl_map_version(&(mmap->mmp_map));
+
+	j = 0;
+	k = 0;
+	oid = md->omd_id;
+	key = oid.lo;
+
+	rc = 0;
+	target = NULL;
+
+	D_ALLOC_ARRAY(dom_used, mmap->dom_used_length);
+	D_ALLOC_ARRAY(used_targets, layout->ol_nr + 1);
+
+	if (dom_used == NULL)
+		return -DER_NOMEM;
+	if (used_targets == NULL)
+		return -DER_NOMEM;
+
+	/**
+	 * If the object class is a special class then the first shard must be
+	 * hand picked because there is no other way to specify a starting
+	 * location.
+	 */
+	if (daos_obj_id2class(oid) == DAOS_OC_R3S_SPEC_RANK ||
+	    daos_obj_id2class(oid) == DAOS_OC_R1S_SPEC_RANK ||
+	    daos_obj_id2class(oid) == DAOS_OC_R2S_SPEC_RANK) {
+
+		rc = mapless_obj_spec_place_get(mmap, oid, &target, dom_used,
+						mmap->dom_used_length);
+
+		if (rc) {
+			D_ERROR("special oid "DF_OID" failed: rc %d\n",
+				DP_OID(oid), rc);
+			return rc;
+		}
+
+		layout->ol_shards[0].po_target = target->ta_comp.co_id;
+		layout->ol_shards[0].po_shard = 0;
+		layout->ol_shards[0].po_fseq = target->ta_comp.co_fseq;
+
+		if (pool_target_unavail(target)) {
+			rc = remap_alloc_one(remap_list, 0, target);
+			if (rc)
+				D_GOTO(out, rc);
+		}
+
+		/** skip the first shard because it's been
+		 * determined by Obj class
+		 */
+		j = 1;
+		k = 1;
+	}
+
+	rc = pool_map_find_domain(mmap->mmp_map.pl_poolmap, PO_COMP_TP_ROOT,
+				  PO_COMP_ID_ALL, &root);
+	if (rc == 0) {
+		D_ERROR("Could not find root node in pool map.");
+		rc = -DER_NONEXIST;
+		goto out;
+	}
+
+	for (i = 0; i < mop->mop_grp_nr; i++) {
```

Comment:
The first shard has been chosen if the SPEC obj class is used, shouldn't we skip it here?

Reply:
We skip it in the inner loop on the next line.

"for (; j < mop->mop_grp_size; j++, k++)"

Where j is set at the end of SPEC handler block.

Reply:
I see, I missed that. Thanks.